src/validation.cpp

```diff
@@ -2359,6 +2360,87 @@ CScript calculate_contract(const CScript& federationRedeemScript, const CScript&
     return scriptDestination;
 }
 
+bool GetAmountFromParentChainPegin(CAmount& amount, const Sidechain::Bitcoin::CTransaction& txBTC, unsigned int nOut)
+{
+    amount = txBTC.vout[nOut].nValue;
+    return true;
+}
+
+bool GetAmountFromParentChainPegin(CAmount& amount, const CTransaction& txBTC, unsigned int nOut)
+{
+    if (!txBTC.vout[nOut].nValue.IsExplicit()) {
+        return false;
+    }
+    amount = txBTC.vout[nOut].nValue.GetAmount();
+    return true;
+}
+
+template<typename T>
+static bool GetBlockAndTxFromMerkleBlock(uint256& block_hash, uint256& tx_hash, T& merkle_block, const std::vector<unsigned char>& merkle_block_raw)
+{
+    try {
+        std::vector<uint256> tx_hashes;
+        std::vector<unsigned int> tx_indices;
+        CDataStream merkle_block_stream(merkle_block_raw, SER_NETWORK, PROTOCOL_VERSION);
+        merkle_block_stream >> merkle_block;
+        block_hash = merkle_block.header.GetHash();
+
+        if (!merkle_block_stream.empty()) {
+           return false;
+        }
+        if (merkle_block.txn.ExtractMatches(tx_hashes, tx_indices) != merkle_block.header.hashMerkleRoot || tx_hashes.size() != 1) {
+            return false;
+        }
+        tx_hash = tx_hashes[0];
+    } catch (std::exception& e) {
+        // Invalid encoding of merkle block
+        return false;
+    }
+    return true;
+}
+
+template<typename T>
+static bool CheckPeginTx(const std::vector<unsigned char>& tx_data, T& pegtx, const COutPoint& prevout, const CAmount claim_amount, const CScript& claim_script)
+{
```

Comment:
- The `T& pegtx` parameter is set and used internally to this function but is ignore by callers, is it needed in the parameter list to *CheckPeginTx()*? It looks like it's just there for type selection for the plumbing underneath? Alternatively you could add a local instantiation before the `try`:

    `T pegtx; // or shared_ptr<const T> pegtx;`

 - Also, because the function can only take *T*'s of type *CTransactionRef*, maybe a more helpful name could be used for the template param (e.g. *"TransactionType"*)?

Reply:
Yes, if the parameters are only used internally they can just be turned into local variables, good point.
CTransactionRef (of both types) uses shared_ptr instead of unique pointer, I assume because other parts need that. I guess I could use a unique_ptr instead of CTransactionRef, but this is what was being done and it seems simpler.

Reply:
oh, no, sorry, I cannot do that, the paremeter is needed to select which one to use within the if/else clause. I know this is hacky and probably can be done in a more elegant way, but I don't see it at this point.